job-server/src/spark.jobserver/JobServerNamedObjects.scala

```diff
@@ -0,0 +1,121 @@
+
+  override def getOrElseCreate[O <: NamedObject](name: String, objGen: => O)
+                                (implicit timeout: Timeout = defaultTimeout,
+                                    persister: NamedObjectPersister[O]): O = {
+
+    val res : Option[O] = getExistingObject(name)
+    res match {
+      case Some(namedObject) =>
+        namedObject
+      case None =>
+        logger.info("Named object [{}] not found, starting creation", name)
+        val f = cachedOp(name, createObject(objGen, name))
+        f.await match {
+          case obj : O => obj
+          case NamedRDD(_, _) => throw new IllegalArgumentException("Incorrect type for named object")
+        }
+    }
+  }
+
+  // we can wrap the operation with caching support
+  // (providing a caching key)
+  private def cachedOp[O <: NamedObject](name: String, f: () => O):
+                           Future[NamedObject] = namesToObjects(name) {
+    f()
+  }
+
+  private def createObject[O <: NamedObject](objGen: => O, name: String)
+               (implicit persister: NamedObjectPersister[O]): () => O = {
+    () =>
+      {
+        val namedObj: O = objGen
+        persister.saveToContext(namedObj, name)
+        namedObj
+      }
+  }
+
+  private def getExistingObject[O](name: String): Option[O] = {
+    //get:  retrieves the future instance that is currently in the cache for the given key.
+    // Returns None if the key has no corresponding cache entry.
+    namesToObjects.get(name) match {
+      case Some(f : Future[O]) =>
+        Some(f.await)
+        //TODO - the implementation for NamedRDDs checked here as follows. Do
+        // we still need to do this?
+      //        sparkContext.getPersistentRDDs.get(rdd.id) match {
+      //          case Some(rdd) => Some(rdd)
+      //          case None =>
+      //            // If this happens, maybe we never knew about this RDD,
+      // or maybe we had a name -> id mapping, but
+      //            // spark's MetadataCleaner has evicted this RDD from
+      //the cache because it was too old, and we need
+      //            // to forget about it. Remove it from our names -> ids map
+      //            // and respond as if we never knew about it.
+      //            namesToRDDs.remove(name)
+      //            None
+      //        }
+      case None =>
+        //appears that we have never seen this named object before or that it was removed
+        None
+    }
+  }
+  override def get(name: String)(implicit timeout: Timeout = defaultTimeout): Option[NamedObject] = {
+    getExistingObject(name) match {
+      case objectOpt: Option[NamedObject] @unchecked => objectOpt //TODO  rddOpt.map { rdd=> refresh(rdd) }
+    }
+  }
+
+  //TODO - isn't this identical to getOrElseCreate?
+  // only that the object may already exist and we need to remove it - unless the old
+  // object (rdd) has the same id TODO - is that even possible? The API does not
+  // mention that special case either
```

Comment:
Also: there's a problem here if we have 2 jobs that use the same name for different type of objects, one will end up with  not only another object but an instance from another type (it will actually throw an exception in the getOrElseCreate